RFS/src/main/java/com/rfs/common/S3Repo.java

```diff
@@ -35,55 +42,61 @@ private static int extractVersion(String key) {
-                .bucket(bucketName)
-                .prefix(prefix)
+                .bucket(s3RepoUri.bucketName)
+                .prefix(s3RepoUri.key)
                 .build();
 
-        ListObjectsV2Response listResponse = s3Client.listObjectsV2(listRequest);
+        ListObjectsV2Response listResponse = s3Client.listObjectsV2(listRequest).join();
 
         Optional<S3Object> highestVersionedIndexFile = listResponse.contents().stream()
                 .filter(s3Object -> s3Object.key().matches(".*index-\\d+$")) // Regex to match index files
                 .max(Comparator.comparingInt(s3Object -> extractVersion(s3Object.key())));
 
-        return highestVersionedIndexFile
-                .map(s3Object -> "s3://" + bucketName + "/" + s3Object.key())
-                .orElse(null);
+        String rawUri = highestVersionedIndexFile
+                .map(s3Object -> "s3://" + s3RepoUri.bucketName + "/" + s3Object.key())
+                .orElse("");
+        return new S3Uri(rawUri);
     }
 
     protected void ensureS3LocalDirectoryExists(Path localPath) throws IOException {
         Files.createDirectories(localPath);
     }
 
-    private void downloadFile(String s3Uri, Path localPath) throws IOException {
-        logger.info("Downloading file from S3: " + s3Uri + " to " + localPath);
+    protected boolean doesFileExistLocally(Path localPath) {
+        return Files.exists(localPath);
+    }
+
+    private void ensureFileExistsLocally(S3Uri s3Uri, Path localPath) throws IOException {
         ensureS3LocalDirectoryExists(localPath.getParent());
 
-        String bucketName = s3Uri.split("/")[2];
-        String key = s3Uri.split(bucketName + "/")[1];
+        if (doesFileExistLocally(localPath)) {
+            logger.debug("File already exists locally: " + localPath);
+            return;
+        }
 
+        logger.info("Downloading file from S3: " + s3Uri.uri + " to " + localPath);
         GetObjectRequest getObjectRequest = GetObjectRequest.builder()
-                .bucket(bucketName)
-                .key(key)
+                .bucket(s3Uri.bucketName)
+                .key(s3Uri.key)
                 .build();
 
-        s3Client.getObject(getObjectRequest, ResponseTransformer.toFile(localPath));
+        s3Client.getObject(getObjectRequest, AsyncResponseTransformer.toFile(localPath)).join();
     }
 
-    public static S3Repo create(Path s3LocalDir, String s3Uri, String s3Region) {
-        S3Client s3Client = S3Client.builder()
-                .region(Region.of(s3Region))
-                .credentialsProvider(DefaultCredentialsProvider.create())
-                .build();
+    public static S3Repo create(Path s3LocalDir, S3Uri s3Uri, String s3Region) {
+        S3AsyncClient s3Client = S3AsyncClient.crtBuilder()
+                                                   .credentialsProvider(DefaultCredentialsProvider.create())
+                                                   .region(Region.of(s3Region))
```

Comment:
Good question; I think the answer here is "no" until there's a clear need for it.  It will help enforce regionalization which, in general, is a good thing.